src/defrag.c

```diff
@@ -1159,17 +1162,33 @@ void activeDefragCycle(void) {
                 db = NULL;
                 server.active_defrag_running = 0;
 
+                /* If the defragmentation hits and defragmented bytes are both
+                 * below 1%, which indicates low defragmentation efficiency.
+                 * Reduce the defragmentation speed by 10% for the next cycle. */
+                long long last_hits = server.stat_active_defrag_hits - start_hits;
+                long long last_misses = server.stat_active_defrag_misses - start_misses;
+                if (last_hits < (last_hits + last_misses) * 0.01 &&
```

Comment:
If the fragmentation bytes fluctuates, and the delta is sometimes positive, sometimes negative, we'll keep resetting the decay factor.
I think it could be better to look at the fragmentation ratio, not bytes, and check that the new one is not higher than the old one (by more than 5%). If it's lower, it's ok, since we have the other condition.

I.e. the main condition is that the hits are low. And the secondary condition is if the fragmentation increases.

Does that make sense to you?

Reply:
feels like I'm on the same way as you.
now reset the speed to 1.0 in the following 3 cases:
1. hits for more than 1%
2. the new fragment bytes is greater than the old fragment bytes, the increment is negative.
3. the new fragment bytes is smaller than the old fragment bytes, the defragmented is greater than >1%, which I think is the same as your 5%.